### PR TITLE
Migrate JS to TS

### PR DESCRIPTION
**Pull request by Juanjofp**
It should works, all test passed and linter too. xDD

### CODE REVIEW

File: __tests__/dependencies.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ describe(
                     mapperService: expect.any(Function)
                 });
                 const depsDb = dependencies(['dbClient']);
-                await depsDb.dbClient.close();
+                await depsDb!.dbClient!.close();
```
Comment (francescorivola) [inline __tests__/dependencies.ts:23]: interesting, I do not have to use the non-null assertion. Maybe I do not have the same typescript setup.

File: __tests__/dependencies.ts [hunk 2]
```diff
@@ -63,26 +63,26 @@ describe(
...
-        it(
-            'receive no params and return an object with no dependencies when dependency not loaded',
-            async () => {
-                const dependencies = await createDependencies(
-                    {
-                        DBNAME: 'CUSTOM_DBNAME',
-                        SOURCES: 'CUSTOM_SOURCES_COLLECTION'
-                    }
-                );
+        // it(
```
Comment (francescorivola) [inline __tests__/dependencies.ts:70]: if not used I would remove this commented code.

File: __tests__/route-mappers.ts [hunk 3]
```diff
@@ -203,17 +201,16 @@ describe('Route mapper', () => {
         const response = await server.inject({
             method: 'POST',
             url: '/mappers/' + sourceId,
-            query: context.params,
+            query: context.params as {[key: string]: string;},
```
Comment (francescorivola) [inline __tests__/route-mappers.ts:204]: why do you need this as?

Comment (francescorivola) [reply __tests__/route-mappers.ts:204]: if the ContextRequest params options match the same type maybe you can avoid the as.

File: src/database.ts [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
         useNewUrlParser: true,
         useUnifiedTopology: true,
         ignoreUndefined: true
     });
 }
 
-export async function getAndSetupDatabase(client, databaseName) {
-    const db = client.db(databaseName);
+export async function getAndSetupDatabase(client: MongoClient, databaseName: string): Promise<Db> {
+    const db: Db = client.db(databaseName);
```
Comment (francescorivola) [inline src/database.ts:21]: maybe Db type can just be inferred. No need to explicit say the type here.

File: src/errors.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if (!ErrorResponseList) {
+        const extendedErrorInfo: ErrorResponse = {...errorResponse, meta: {...errorInfo.meta, ...meta}};
+        return { errors: [extendedErrorInfo]};
+    }
+    ErrorResponseList.errors.push({...errorResponse, meta});
+    return ErrorResponseList;
+};
+
+export const encodeErrorFromType = (ErrorResponseList: ErrorResponseList = {errors: []}, type: symbol, meta: object = {}): ErrorResponseList => {
+    const extendedErrorInfo: ErrorInfo = findErrorByType(type);
```
Comment (francescorivola) [inline src/errors.ts:89]: maybe : ErrorInfo can be inferred by the return type of the function.

File: src/errors.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    return encodeErrorFromErrorInfo(ErrorResponseList, extendedErrorInfo, meta);
+};
+
+export const encodeErrorFromError = (ErrorResponseList: ErrorResponseList, error: ErrorWithType, meta: object = {}) : ErrorResponseList => {
+    const errorInfo : ErrorInfo = findError(error);
+    return encodeErrorFromErrorInfo(ErrorResponseList, errorInfo, meta);
+};
+
+export const reThrowError = (errorType: symbol, error: ErrorWithType): never => {
```
Comment (francescorivola) [inline src/errors.ts:98]: hehehe, never is very cool :)

File: src/errors.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    throw error;
+};
+
+export const throwError = (errorType: symbol, message: string): never => {
+    const error: ErrorWithType = new Error(message);
+    error.errorType = errorType;
+    throw error;
+};
+
+export const typeOfError = (error: ErrorWithType, errorType: symbol) : boolean => {
```
Comment (francescorivola) [inline src/errors.ts:109]: I would personally prefer name this function isTypeOfError

File: src/routes/admin/sources.ts [hunk 8]
```diff
@@ -65,25 +69,24 @@ const SourceSchema = {
...
+export function buildAdminSourcesRoutes(deps: DependenciesLoader) {
     const {
         sourcesService,
         mappingsService,
         targetsService,
         responsesService
     } = deps(['sourcesService', 'mappingsService', 'targetsService', 'responsesService']);
 
-    async function listSources(request, reply) {
+    async function listSources(request: FastifyRequest, reply: FastifyReply<ServerResponse>) {
```
Comment (francescorivola) [inline src/routes/admin/sources.ts:80]: I must review scheduler code. I think I am not using this Fastify types. Good point.

File: src/routes/admin/sources.ts [hunk 9]
```diff
@@ -92,21 +95,20 @@ export function buildAdminSourcesRoutes(deps) {
         }
     }
 
-    async function findByIdSource(request, reply) {
-        const sourceId = request.params.sourceId;
+    async function findByIdSource(request: FastifyRequest, reply: FastifyReply<ServerResponse>) {
+        const sourceId: string = request.params.sourceId;
```
Comment (francescorivola) [inline src/routes/admin/sources.ts:99]: is explicit type here needed?

File: src/routes/admin/sources.ts [hunk 10]
```diff
@@ -117,32 +119,32 @@ export function buildAdminSourcesRoutes(deps) {
...
-            const invalidParams = [];
+            const invalidParams: object[] = [];
             const flows = body.flows;
             if (flows) {
                 if (Array.isArray(flows)) {
                     // Check each flow
                     for (const flow of flows) {
                         const targetId = flow.targetId;
-                        const target = await targetsService.getTargetById(targetId);
+                        const target = await targetsService!.getTargetById(targetId);
```
Comment (francescorivola) [inline src/routes/admin/sources.ts:143]: I am not sure why you need to always use the exclamation mark any time you need to get access to service functions.

File: src/server.ts [hunk 11]
```diff
@@ -74,16 +64,15 @@ export function buildServer(deps) {
 	});
 
 	app.setErrorHandler((error, request, reply) => {
-		if (error.statusCode < 500) {
+		if (error.statusCode && error.statusCode < 500) {
```
Comment (francescorivola) [inline src/server.ts:67]: this && sucks as hit our code coverage xD

File: tests-utils/mock-requsts.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import nock from 'nock';
-
-export const createGetRequest = (host, path, headers, status, response, respHeaders) => {
+import type { DefaultHeaders } from 'fastify';
```
Comment (francescorivola) [inline tests-utils/mock-requsts.ts:2]: typo in file name mock-requsts.ts

File: src/utils/logger.ts [hunk 13]
```diff
@@ -0,0 +1 @@
+export const logger = (...args: any[]) => {};//console.log((new Date()).toISOString(), ...args);
```
Comment (francescorivola) [inline src/utils/logger.ts:1]: Should be this the pino logger? Did you have disabled temporally?

File: __tests__/route-mappers.ts (hunk 3 above)
Comment (Juanjofp) [reply __tests__/route-mappers.ts:204]: because this line was write before its type... xDDD

File: src/dependencies.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+import type { MongoClient, Db, Collection } from 'mongodb';
+
+export type DependenciesConfig = {
+    DBNAME?: string;
+    SOURCES?: string;
+    MAPPINGS?: string;
+    TARGETS?: string;
+    RESPONSES?: string;
+};
+export type DependenciesName = 'config'|'dbClient'|'db'|'sourcesCollection'|'sourcesService'|'mappingsCollection'|'mappingsService'|'targetsCollection'|'targetsService'
```
Comment (francescorivola) [inline src/dependencies.ts:26]: I meant this could be
`export type DependenciesName = keyof Dependencies;`

Comment (Juanjofp) [reply src/dependencies.ts:26]: Great!

File: tests-utils/mock-requsts.ts (hunk 12 above)
Comment (Juanjofp) [reply tests-utils/mock-requsts.ts:2]: I forgot this one! I Will fix It in master

**Pull request merged**